kernel/src/acpi/madt.rs

```diff
@@ -0,0 +1,432 @@
+use alloc::vec::Vec;
+
+use crate::{
+    memory::{
+        reinterpret_memory, DynamicallySized, DynamicallySizedItem, DynamicallySizedObjectIterator,
+        Validateable,
+    },
+    println,
+};
+use core::mem::size_of;
+
+use super::AcpiTableHandle;
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+struct MadtEntryHeader {
+    entry_type: u8,
+    length: u8,
+}
+
+impl DynamicallySized for MadtEntryHeader {
+    fn size(&self) -> usize {
+        self.length as usize
+    }
+}
+
+const MAX_ENTRY_SIZE: u8 = 128;
+
+impl Validateable for MadtEntryHeader {
+    fn validate(&self) -> bool {
+        self.length >= 2 && self.length <= MAX_ENTRY_SIZE
+    }
+}
+
+// It doesn't really make sense to use an enum for the type, since we don't know all the possible values (since the spec may change).
+const MADT_ENTRY_TYPE_LOCAL_APIC: u8 = 0;
+const MADT_ENTRY_TYPE_IO_APIC: u8 = 1;
+const MADT_ENTRY_TYPE_INTERRUPT_SOURCE_OVERRIDE: u8 = 2;
+const MADT_ENTRY_TYPE_NON_MASKABLE_INTERRUPT_SOURCE: u8 = 3;
+const MADT_ENTRY_TYPE_LOCAL_APIC_NMI: u8 = 4;
+const MADT_ENTRY_TYPE_LOCAL_APIC_ADDRESS_OVERRIDE: u8 = 5;
+const MADT_ENTRY_TYPE_IO_SAPIC: u8 = 6;
+const MADT_ENTRY_TYPE_LOCAL_SAPIC: u8 = 7;
+const MADT_TYPE_INTERRUPT_SOURCES: u8 = 8;
+const MADT_TYPE_LOCAL_X2APIC: u8 = 9;
+const MADT_TYPE_LOCAL_X2APIC_NMI: u8 = 0xa;
+const MADT_TYPE_GENERIC_INTERRUPT_CONTROLLER_CPU_INTERFACE: u8 = 0xb;
+const MADT_TYPE_GENERIC_INTERRUPT_CONTROLLER_DISTRIBUTOR: u8 = 0xc;
+const MADT_TYPE_GENERIC_INTERRUPT_CONTROLLER_MSI_FRAME: u8 = 0xd;
+const MADT_TYPE_GENERIC_INTERRUPT_CONTROLLER_REDISTRIBUTOR: u8 = 0xe;
+const MADT_TYPE_GENERIC_INTERRUPT_CONTROLLER_TRANSLATION_SERVICE: u8 = 0xf;
+
+macro_rules! default_operator {
+    ($left:expr, $default:tt|$given:tt, $right:expr) => {
+        $left $given $right
+    };
+    ($left:expr, $default:tt, $right:expr) => {
+        $left $default $right
+    };
+}
+
+macro_rules! madt_entry {
```

Comment:
I'm not sure that these macro rules are making the code more readable. Is it possible to do this size comparison without the custom macro?

Reply:
Yes, but it would be exceedingly repetative. Every single one would have to have the structure itself, plus an impl block for the getter methods, plus another impl block for Validatable. But is it just the => true but that you don't like? Because we could easily just remove it. I added that part without really considering that there would probably be no places where we could reasonably validate anything. Without that bit it would just be a simple Rust structure definition wrapped in the macro, possibly with the comparison operator attribute. I suppose we could remove that as well and just not use the macro when we need to have a different operator as well.